src/data_store/generic_data_store.cpp

```diff
@@ -293,4 +311,57 @@ void generic_data_store::init_minibatch() {
   }
 }
 
+std::pair<std::string, std::string> generic_data_store::get_pathname_and_prefix(std::string s) {
+  int num_slash = std::count(s.begin(), s.end(), '/');
+  if (num_slash < 1 || s.back() == '/') {
+    std::stringstream err;
+    err << __FILE__ << " " << __LINE__ << " :: "
+        << "<string> must be of the form: <pathname>/prefix";
+    throw lbann_exception(err.str());
+  }
+
+  size_t j = s.rfind('/');
+  std::string prefix = s.substr(j+1);
+  std::string pathname = s.substr(0, j);
+  return std::make_pair(prefix, pathname);
+}
+
+void generic_data_store::create_dirs(std::string s) {
+  if (s.back() != '/') {
+    s += '/';
+  }
+  size_t i = s.find('/', 1);
+  while (i != std::string::npos) {
+    std::string s2 = s.substr(0, i);
+    const int dir_err = mkdir(s2.c_str(), S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);
+    if (dir_err == -1 && errno != 17) { // 17: File Exists
+      std::stringstream err;
+      err << __FILE__ << " " << __LINE__ << " :: "
+          << "failed to create directory: " << s2 << "\n"
+          << "error code is: " << errno << " -> " << std::strerror(errno)
+          << "\n" << getenv("SLURMD_NODENAME");
+      throw lbann_exception(err.str());
+    }
+    i = s.find('/', i+1);
+  }
+}
+
+std::string generic_data_store::run_cmd(std::string cmd, bool exit_on_error) {
+  std::array<char, 128> buffer;
+  std::string result;
+  std::shared_ptr<FILE> pipe(popen(cmd.c_str(), "r"), pclose);
```

Comment:
If you have already tested this on ray, then disregard this comment. I had some issue with passing string::c_str() to call a system function on ray. For some reason, I had to use the real c-string (null-terminated) in a character array.

Reply:
I've only worked on catalyst. " I had some issue with passing string::c_str() to call a system function on ray ..." Thanks for the heads-up; and that seems weird.

Reply:
modified to copy cmd to c-style string